tests/utilities_t/executors_t/test_sshexecutor.py

```diff
@@ -10,18 +10,63 @@
 
 import asyncio
 import yaml
+import contextlib
+from asyncstdlib import contextmanager as asynccontextmanager
+
+
+DEFAULT_MAX_SESSIONS = 10
 
 
 class MockConnection(object):
-    def __init__(self, exception=None, **kwargs):
+    def __init__(self, exception=None, __max_sessions=DEFAULT_MAX_SESSIONS, **kwargs):
         self.exception = exception and exception(**kwargs)
+        self.max_sessions = __max_sessions
+        self.current_sessions = 0
+
+    @contextlib.contextmanager
+    def _multiplex_session(self):
+        if self.current_sessions >= self.max_sessions:
+            raise ChannelOpenError(code=2, reason="open failed")
+        self.current_sessions += 1
+        try:
+            yield
+        finally:
+            self.current_sessions -= 1
 
     async def run(self, command, input=None, **kwargs):
-        if self.exception:
-            raise self.exception
-        return AttributeDict(
-            stdout=input and input.decode(), stderr="TestError", exit_status=0
-        )
+        with self._multiplex_session():
+            if self.exception:
+                raise self.exception
+            if command.startswith("sleep"):
+                _, duration = command.split()
+                await asyncio.sleep(float(duration))
+            elif command != "Test":
+                raise ValueError(f"Unsupported mock command: {command}")
+            return AttributeDict(
+                stdout=input and input.decode(), stderr="TestError", exit_status=0
+            )
+
+    async def create_process(self):
+        @asynccontextmanager
+        async def fake_process():
+            with self._multiplex_session():
+                yield
+
+        return fake_process()
+
+
+class TestSSHExecutorUtilities(TestCase):
+    def test_max_sessions(self):
+        with self.subTest(sessions="default"):
+            self.assertEqual(
+                DEFAULT_MAX_SESSIONS, run_async(probe_max_session, MockConnection())
```

Comment:
What is idea behind this test?
Isn't it automatically successful? 
Or am I misunderstanding something here?

Reply:
It's a sanity test that the Mock/probe behaves as expected. Other tests rely on the default as well, so if they fail this test reveals whether the issue is with the default or the other tests.

Reply:
It test the functionality of the `probe_max_session` function defined in https://github.com/MatterMiners/tardis/blob/2d31c693b6090d0fcb0d5219c3593403fc63e67e/tardis/utilities/executors/sshexecutor.py#L15-L31. In case someone changes the code so, that the functionality is not obtained anymore. The test will fail. For example someone removes https://github.com/MatterMiners/tardis/blob/2d31c693b6090d0fcb0d5219c3593403fc63e67e/tardis/utilities/executors/sshexecutor.py#L28.

Reply:
Okay thanks. That makes perfectly sense!